pygen/pygen_src/isa/riscv_compressed_instr.py

```diff
@@ -0,0 +1,228 @@
+            self.imm_len = 6
+        elif self.format in [riscv_instr_format_t.CL_FORMAT, riscv_instr_format_t.CS_FORMAT]:
+            self.imm_len = 5
+        elif self.format in [riscv_instr_format_t.CJ_FORMAT]:
+            self.imm_len = 11
+        elif self.format in [riscv_instr_format_t.CB_FORMAT]:
+            if self.instr_name == riscv_instr_name_t.C_ANDI:
+                self.imm_len = 6
+            else:
+                self.imm_len = 7
+        elif self.format in [riscv_instr_format_t.CB_FORMAT, riscv_instr_format_t.CIW_FORMAT]:
+            self.imm_len = 8
+        if self.instr_name in [riscv_instr_name_t.C_SQ, riscv_instr_name_t.C_LQ,
+                               riscv_instr_name_t.C_LQSP, riscv_instr_name_t.C_SQSP,
+                               riscv_instr_name_t.C_ADDI16SP]:
+            self.imm_align = 4
+        elif self.instr_name in [riscv_instr_name_t.C_SD, riscv_instr_name_t.C_LD,
+                                 riscv_instr_name_t.C_LDSP, riscv_instr_name_t.C_SDSP]:
+            self.imm_align = 3
+        elif self.instr_name in [riscv_instr_name_t.C_SW, riscv_instr_name_t.C_LW,
+                                 riscv_instr_name_t.C_LWSP, riscv_instr_name_t.C_SWSP,
+                                 riscv_instr_name_t.C_ADDI4SPN]:
+            self.imm_align = 2
+        elif self.instr_name in [riscv_instr_name_t.C_LUI]:
+            self.imm_align = 12
+        elif self.instr_name in [riscv_instr_name_t.C_J, riscv_instr_name_t.C_JAL,
+                                 riscv_instr_name_t.C_BNEZ, riscv_instr_name_t.C_BEQZ]:
+            self.imm_align = 1
+
+    def extend_imm(self):
+        if self.instr_name != riscv_instr_name_t.C_LUI:
+            super().extend_imm()
+            self.imm = self.imm << self.imm_align
+
+    def set_rand_mode(self):
+        if self.format in [riscv_instr_format_t.CR_FORMAT]:
+            if self.category in [riscv_instr_category_t.JUMP]:
+                self.has_rd = 0
+            else:
+                self.has_rs1 = 0
+                self.has_imm = 0
+        elif self.format in [riscv_instr_format_t.CI_FORMAT, riscv_instr_format_t.CIW_FORMAT]:
+            self.has_rs2 = 0
+            self.has_rs1 = 0
+        elif self.format in [riscv_instr_format_t.CSS_FORMAT]:
+            self.has_rs1 = 0
+            self.has_rd = 0
+        elif self.format in [riscv_instr_format_t.CL_FORMAT]:
+            self.has_rs2 = 0
+        elif self.format in [riscv_instr_format_t.CS_FORMAT]:
+            self.has_rd = 0
+        elif self.format in [riscv_instr_format_t.CA_FORMAT]:
+            self.has_rs1 = 0
+            self.has_imm = 0
+        elif self.format == riscv_instr_format_t.CJ_FORMAT:
+            self.has_rs1 = 0
+            self.has_rs2 = 0
+            self.has_rd = 0
+        elif self.format == riscv_instr_format_t.CB_FORMAT:
+            if self.instr_name != riscv_instr_name_t.C_ANDI:
+                self.has_rd = 0
+                self.has_rs2 = 0
```

Comment:
similar here, I think `has_rs2` should be set to zero independently of the `if...else` block